improve usability of pulumi yaml project file (#1821)
pulumi/pulumi-hugo: themes/default/content/docs/reference/pulumi-yaml.md

options:
  refresh: true
template:
  description: An example template
  config:
    description: An example template config
    default: An example default
    secret: true
plugins:
  providers:
    name: aws
    path: path-to-aws-provider/
    version: 5.10.0
```

## Attributes

| Name | Required | Description | 
| - | - | - | - |
| `name` | required | The name of the project containing alphanumberic characters, hyphens, underscores, and period. |
| `runtime` | required | The installed language runtime of the project: `nodejs`, `python`, `go`, `dotnet`, `java` or `yaml` to use. | 
| `description` | optional | The description of the project. |
| `main` | optional | Overide of the working directity default of the location of the Pulumi program. |
| `stackConfigDir` | optional | Config directory location relative to the location of `Pulumi.yaml` | 
| `backend` | optional | The [backend]({{< relref "/docs/intro/concepts/state" >}}) of the project. |
| `options` | optional | Additional project options. |
| `template` | optional | Config to be used when creatinig new stacks in the project. |
| `plugins` | optional | Override the plugin selection. |

### `runtime` options

> The runtime attribute has an additional optional `options` option where you can specify additional runtime configuration.

Author: i know `optional options option` is ridic, what can i use instead?

Reviewer: How about:

> The runtime attribute has an additional property called `options` where you can further specify runtime configuration.

| `binary` | Applicable for the go, .net, and java runtimes | Path to pre-built executable. |
| `virtualenv` | Ony applicable fo rthe python runtime | Virtual environment path. | 
| `compiler` | Only applicable to YAML projects | Executable and arguments that emit to standard out. |

#### About `binary`

 - For Go, if not specified, go sources in $CWD will be invoked via `go run`.
 - For .NET, f not specified, a .NET project in $CWD will be invoked via `dotnet run`.

#### About `virtualenv`

New Python projects created with `pulumi new` have this option set by default. If not specified, Pulumi will invoke the `python3` command it finds on $PATH (falling back to `python`) to run the Python program. If you'd like to use a virtual environment without the `virtualenv` option, you'll need to run any `pulumi` commands (such as `pulumi up`) from an activated virtual environment shell (or, if using a tool like [Pipenv](https://github.com/pypa/pipenv), prefix any `pulumi` commands with `pipenv run pulumi ...`).

### `options` options

| Name | Description | Default |
| - | - | - |
| `refresh` | Boolean indicating whether to refresh the state before performing a Pulumi operation | `true` |

> ### `template` options

Author: i need help understanding this

    weight: 5
---

The `Pulumi.yaml` project file specifies metadata about your project, such as the project name and language runtime for your project.

## Example project file

```
name: Example Pulumi project file
runtime: python
description: An example project
main: example-project/
stackConfigDir: config/
backend:
  url: https://pulumi.example.com
options:
  refresh: true
template:
  description: An example template
  config:
    aws:region:
      description: The AWS region to deploy into
      default: us-east-1
      secret: true
plugins:
  providers:
    name: aws
    path: ../.../bin
  languages:
    name: yaml
    path: ../../../pulumi-yaml/bin
    version: 1.2.3
```

## Attributes

| Name | Required | Description |
| - | - | - |

> | `name` | required | Name of the project containing alphanumberic characters, hyphens, underscores, and period. |

Author: ```suggestion
| `name` | required | Name of the project containing alphanumeric characters, hyphens, underscores, and periods. |
```

| `runtime` | required | Installed language runtime of the project: `nodejs`, `python`, `go`, `dotnet`, `java` or `yaml`. |
| `description` | optional | Description of the project. |
| `main` | optional | Path to the Pulumi program. The default is the working directory. |
| `stackConfigDir` | optional | Config directory location relative to the location of `Pulumi.yaml`. |
| `backend` | optional | [Backend]({{< relref "/docs/intro/concepts/state" >}}) of the project. |
| `options` | optional | Additional project options. |
| `template` | optional | Config to be used when creating new stacks in the project. |
| `plugins` | optional | Override for the plugin selection. Intended for use in developing pulumi plugins.  |

### `runtime` options

The runtime attribute has an additional optional `options` option where you can specify additional runtime configuration.

| Name | Use case | Description |
| - | - | - |
| `typescript` | Only applicable for the nodejs runtime | Boolean indicating whether to use `ts-node` or not. |
| `binary` | Applicable for the go, .net, and java runtimes | Path to pre-built executable. |

> | `virtualenv` | Ony applicable fo the python runtime | Virtual environment path. |

Author: ```suggestion
| `virtualenv` | Ony applicable for the python runtime | Virtual environment path. |
```

> | `compiler` | Only applicable to YAML projects | Executable and arguments that emit to standard out. |

Author: ```suggestion
| `compiler` | Only applicable for YAML projects | Executable and arguments that emit to standard out. |
```

### `options` options

| Name | Required | Description | Default |
| - | - | - |
| `refresh` | optional | Boolean indicating whether to refresh the state before performing a Pulumi operation. | `true` |

### `template` options

| Name | Required | Description |
| - | - | - |
| `description` | optional | Description of the template. |
| `config` | required | Config to apply to each stack in the project. |

#### `config` options

| Name | Required | Description |
| - | - | - |
| `description` | optional | Description of the config. |
| `default` | optional | Default value of the config. |
| `secret` | optional | Boolean indicating if the configuration is labeled as a secret. |

### `plugins` options

Use this option to link to local plugin binaries. This option is intended for use in developing pulumi plugins.

| Name | Required | Description |
| `providers` | optional | something |
| `analyzers` | optional | something |
| `languages` | optional | something |

#### Options for `providers`, `analyzers`, and `languages`

| Name | Required | Description |
| - | - | - |
| `name` | required | Name of the plugin. |
| `path` | optional | Path to the plugin folder. |

> | `version` | optional | Version of the plugin, if not set this will match for any version the engine requests. |

Author: ```suggestion
| `version` | optional | Version of the plugin, if not set, will match any version the engine requests. |
```

    weight: 5
---

The `Pulumi.yaml` project file specifies metadata about your project, such as the project name and language runtime for your project.

> ## Example project files

Reviewer 2: If we're not Using Title Case for Headings, we should indicate this in the Style Guide. (I don't have a strong preference either way - just that we are consistent.) It looks like our existing docs do use proper title case.

Author: oh lol, ill just align in this pr but we should revisit this

Author: https://github.com/pulumi/pulumi-hugo/pull/1834

Reviewer 2: Makes this consistent with the wording in the Registry:

```suggestion
## Example Usage
```

Author: @jkodroff hmm, id rather this be specific to reduce confusion then align with the docs in the registry. am i missing something about why this needs to align with registry?

Author: also, would you be totally against me keeping the headings as is and me going to fix up the rest, or would you rather these be title case for now?

Reviewer 2: Doesn't _need_ to align with the registry, but I thought it read better and was consistent to boot. No hard feelings if you feel differently.

Not totally against keeping the headings as-is since this PR is such a significant improvement over what came before.

### Example project file with only required attributes

```
name: Example Pulumi project file with only requires attributes
runtime: nodejs
```

### Example project file with all possible attributes

> ```

Reviewer 2: ```suggestion
```yaml
```

name: Example Pulumi project file will all possible attributes
runtime: python
description: An example project
main: example-project/
stackConfigDir: config/
backend:
  url: https://pulumi.example.com
options:
  refresh: true
template:
  description: An example template
  config:
    aws:region:
      description: The AWS region to deploy into
      default: us-east-1
      secret: true
plugins:
  providers:
    name: aws
    path: ../.../bin
  languages:
    name: yaml
    path: ../../../pulumi-yaml/bin
    version: 1.2.3
```

## Attributes

> | Name | Required | Description |

Reviewer 2: 🙏 THANK YOU.

| - | - | - |
| `name` | required | Name of the project containing alphanumeric characters, hyphens, underscores, and periods. |
| `runtime` | required | Installed language runtime of the project: `nodejs`, `python`, `go`, `dotnet`, `java` or `yaml`. |
| `description` | optional | Description of the project. |
| `main` | optional | Path to the Pulumi program. The default is the working directory. |
| `stackConfigDir` | optional | Config directory location relative to the location of `Pulumi.yaml`. |
| `backend` | optional | [Backend]({{< relref "/docs/intro/concepts/state" >}}) of the project. |
| `options` | optional | Additional project options. |
| `template` | optional | Config to be used when creating new stacks in the project. |
| `plugins` | optional | Override for the plugin selection. Intended for use in developing pulumi plugins.  |

### `runtime` options

> The runtime attribute has an additional optional `options` option where you can specify additional runtime configuration.

Author: ```suggestion
The runtime attribute has an additional property called options where you can further specify runtime configuration.
```

| Name | Use case | Description |
| - | - | - |
| `typescript` | Only applicable for the nodejs runtime | Boolean indicating whether to use `ts-node` or not. |
| `binary` | Applicable for the go, .net, and java runtimes | Path to pre-built executable. |
| `virtualenv` | Ony applicable for the python runtime | Virtual environment path. |
| `compiler` | Only applicable for YAML projects | Executable and arguments that emit to standard out. |

#### About `binary`

> - For Go, if not specified, go sources in $CWD will be invoked via `go run`.

Reviewer 2: ```suggestion
- For Go, if not specified, go sources in `$CWD` will be invoked via `go run`.
```

> - For .NET, f not specified, a .NET project in $CWD will be invoked via `dotnet run`.

Reviewer 2: ```suggestion
- For .NET, if not specified, a .NET project in `$CWD` will be invoked via `dotnet run`.
```

#### About `virtualenv`

> New Python projects created with `pulumi new` have this option set by default. If not specified, Pulumi will invoke the `python3` command it finds on $PATH (falling back to `python`) to run the Python program. To use a virtual environment without the `virtualenv` option, run `pulumi` commands (such as `pulumi up`) from an activated virtual environment shell. Or, if using a tool like [Pipenv](https://github.com/pypa/pipenv), prefix `pulumi` commands with `pipenv run pulumi ...`.

Reviewer 2: ```suggestion
New Python projects created with `pulumi new` have this option set by default. If not specified, Pulumi will invoke the `python3` command it finds on `$PATH` (falling back to `python`) to run the Python program. To use a virtual environment without the `virtualenv` option, run `pulumi` commands (such as `pulumi up`) from an activated virtual environment shell. Or, if using a tool like [Pipenv](https://github.com/pypa/pipenv), prefix `pulumi` commands with `pipenv run pulumi ...`.
```

### `options` options

> | Name | Required | Description | Default |

Reviewer 2: Debatable whether a table is more readable when there's a single line. (Not sure myself.) But I get its inclusion for consistency.

Author: gonna leave as is, for now

        - `default`: (optional) the default value of the config setting - which will be presented to the user as a default.
        - `secret`: (optional) if `true` indicates that this configration value should be marked as secret.

> ### `plugins` options

Author: i need help understanding this
